TypeScript error (Property is not assignable to 'string' index type) in generated output

### Description

I've recently started migrating https://github.com/paambaati/neon-js-sdk from [`openapi-typescript-codegen`](https://www.npmjs.com/package/openapi-typescript-codegen) to `@hey-api/openapi-ts`, and I've run into an error in the generated output.

```
generated/types.gen.ts(1751,5): error TS2411: Property 'pg_settings' of type 'PgSettingsData' is not assignable to 'string' index type 'string'.
generated/types.gen.ts(1752,5): error TS2411: Property 'pgbouncer_settings' of type 'PgbouncerSettingsData' is not assignable to 'string' index type 'string'.
generated/types.gen.ts(1759,5): error TS2411: Property 'autoscaling_limit_min_cu' of type 'number' is not assignable to 'string' index type 'string'.
generated/types.gen.ts(1765,5): error TS2411: Property 'autoscaling_limit_max_cu' of type 'number' is not assignable to 'string' index type 'string'.
generated/types.gen.ts(1766,5): error TS2411: Property 'suspend_timeout_seconds' of type 'number' is not assignable to 'string' index type 'string'.
```

<img width="1053" alt="Screenshot 2024-04-23 at 1 10 53 PM" src="https://github.com/hey-api/openapi-ts/assets/5381764/da331c0f-40e5-47bc-a5c8-31d85f4cb290">


Consider this snippet of code where the TS errors are –

```typescript
export type DefaultEndpointSettings = {
    pg_settings?: PgSettingsData;
    pgbouncer_settings?: PgbouncerSettingsData;
    /**
     * The minimum number of Compute Units. The minimum value is `0.25`.
     * See [Compute size and Autoscaling configuration](https://neon.tech/docs/manage/endpoints#compute-size-and-autoscaling-configuration)
     * for more information.
     *
     */
    autoscaling_limit_min_cu?: ComputeUnit;
    /**
     * The maximum number of Compute Units. See [Compute size and Autoscaling configuration](https://neon.tech/docs/manage/endpoints#compute-size-and-autoscaling-configuration)
     * for more information.
     *
     */
    autoscaling_limit_max_cu?: ComputeUnit;
    suspend_timeout_seconds?: SuspendTimeoutSeconds;
    [key: string]: (string) | undefined;
};
```

One of the types are –

```typescript
/**
 * A raw representation of Postgres settings
 */
export type PgSettingsData = {
    [key: string]: (string);
};
```

In contrast, the types generated by `openapi-typescript-codegen` do not have this problem.

### OpenAPI specification (optional)

https://dfv3qgd2ykmrx.cloudfront.net/api_spec/release/v2.json

### Configuration

```
openapi-ts --name NeonClient --client fetch --input https://dfv3qgd2ykmrx.cloudfront.net/api_spec/release/v2.json --output ./generated
```

### System information (optional)

OS: Darwin MacBook-Pro.local 23.4.0 Darwin Kernel Version 23.4.0: Fri Mar 15 00:12:25 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T6030 arm64
Package version: 0.42.1
Node.js version: 21.7.3

--- a/.changeset/light-ducks-cry.md
+++ b/.changeset/light-ducks-cry.md
@@ -1,0 +1,5 @@
+---
+"@hey-api/openapi-ts": patch
+---
+
+fix: handle additional properties union

--- a/examples/openapi-ts-fetch/src/client/types.gen.ts
+++ b/examples/openapi-ts-fetch/src/client/types.gen.ts
@@ -12,6 +12,11 @@ export type Order = {
   complete?: boolean;
 };
 
+/**
+ * Order Status
+ */
+export type status = 'placed' | 'approved' | 'delivered';
+
 export type Customer = {
   id?: number;
   username?: string;

--- a/packages/openapi-ts/src/openApi/v3/parser/getModelProperties.ts
+++ b/packages/openapi-ts/src/openApi/v3/parser/getModelProperties.ts
@@ -1,4 +1,5 @@
 import { escapeName } from '../../../utils/escape';
+import { unique } from '../../../utils/unique';
 import type { Model } from '../../common/interfaces/client';
 import { getDefault } from '../../common/parser/getDefault';
 import { getPattern } from '../../common/parser/getPattern';
@@ -38,6 +39,17 @@ export const getAdditionalPropertiesModel = (
   }
 
   if (definition.additionalProperties && definition.properties) {
+    const additionalPropertiesType =
+      typeof definition.additionalProperties === 'object' &&
+      definition.additionalProperties.type &&
+      !Array.isArray(definition.additionalProperties.type)
+        ? definition.additionalProperties.type
+        : apModel.base;
+    const additionalProperties = [
+      additionalPropertiesType,
+      ...model.properties.map((property) => property.base),
+    ];
+    apModel.base = additionalProperties.filter(unique).join(' | ');
     apModel.default = getDefault(definition, model);
     apModel.export = 'generic';
     apModel.isRequired = definition.additionalProperties === true;

--- a/packages/openapi-ts/test/sample.cjs
+++ b/packages/openapi-ts/test/sample.cjs
@@ -7,10 +7,19 @@ const main = async () => {
     enums: 'javascript',
     input: './test/spec/v3.json',
     output: './test/generated/v3/',
+    schemas: {
+      export: false,
+    },
+    services: {
+      export: false,
+    },
+    types: {
+      include: '^DictionaryWithPropertiesAndAdditionalProperties',
+    },
   };
 
   const { createClient } = await import(
-    path.resolve(process.cwd(), 'dist/index.js')
+    path.resolve(process.cwd(), 'dist/node/index.cjs')
   );
   await createClient(config);
 };

--- a/packages/openapi-ts/test/spec/v3.json
+++ b/packages/openapi-ts/test/spec/v3.json
@@ -1867,7 +1867,10 @@
         "type": "object",
         "properties": {
           "foo": {
-            "type": "string"
+            "type": "number"
+          },
+          "bar": {
+            "type": "boolean"
           }
         },
         "additionalProperties": {

